sdk/src/openbook/openbookV2Subscriber.ts

```diff
@@ -0,0 +1,169 @@
+	client: OpenBookV2Client;
+
+	public constructor(config: OpenbookV2SubscriberConfig) {
+		this.connection = config.connection;
+		this.programId = config.programId;
+		this.marketAddress = config.marketAddress;
+		this.subscribed = false;
+		if (config.accountSubscription.type === 'polling') {
+			this.subscriptionType = 'polling';
+			this.accountLoader = config.accountSubscription.accountLoader;
+		} else {
+			this.subscriptionType = 'websocket';
+		}
+	}
+
+	public async subscribe(): Promise<void> {
+		if (this.subscribed === true) {
+			return;
+		}
+
+		const anchorProvider = new AnchorProvider(
+			this.connection,
+			new Wallet(Keypair.generate()),
+			{}
+		);
+		const openbookV2Program = new Program(
+			openbookV2Idl as Idl,
+			this.programId,
+			anchorProvider
+		);
+		this.client = new OpenBookV2Client(anchorProvider);
+		const market = await Market.load(this.client, this.marketAddress);
+		this.market = await market.loadOrderBook();
+
+		if (this.subscriptionType === 'websocket') {
+			this.marketCallbackId = this.connection.onAccountChange(
+				this.marketAddress,
+				(accountInfo, _) => {
+					const marketRaw = openbookV2Program.coder.accounts.decode(
+						'Market',
+						accountInfo.data
+					);
+					this.market = new Market(this.client, this.marketAddress, marketRaw);
+					(async () => {
+						await this.market.loadOrderBook();
+					})();
+				}
+			);
+		} else {
+			this.marketCallbackId = await this.accountLoader.addAccount(
+				this.marketAddress,
+				(buffer, _) => {
+					const marketRaw = openbookV2Program.coder.accounts.decode(
+						'Market',
+						buffer
+					);
+					this.market = new Market(this.client, this.marketAddress, marketRaw);
+					(async () => {
+						await this.market.loadOrderBook();
+					})();
+				}
+			);
+		}
+
+		this.subscribed = true;
+	}
+
+	public async getBestBid(): Promise<BN | undefined> {
+		const bids = await this.market.loadBids();
+		const bestBid = bids.best();
+
+		if (bestBid === undefined) {
+			return undefined;
+		}
+
+		return new BN(Math.floor(bestBid.price * PRICE_PRECISION.toNumber()));
+	}
+
+	public async getBestAsk(): Promise<BN | undefined> {
+		const asks = await this.market.loadAsks();
+		const bestAsk = asks.best();
+
+		if (bestAsk === undefined) {
+			return undefined;
+		}
+
+		return new BN(Math.floor(bestAsk.price * PRICE_PRECISION.toNumber()));
+	}
+
+	public getL2Bids(): Generator<L2Level> {
+		return this.getL2Levels('bids');
+	}
+
+	public getL2Asks(): Generator<L2Level> {
+		return this.getL2Levels('asks');
+	}
+
+	*getL2Levels(side: 'bids' | 'asks'): Generator<L2Level> {
+		const basePrecision = Math.ceil(
+			1 / this.market.baseNativeFactor.toNumber()
+		);
+		const pricePrecision = PRICE_PRECISION.toNumber();
+
+		const levels = side === 'bids' ? this.market.bids : this.market.asks;
+
+		for (const order of levels.items()) {
+			const size = new BN(order.size * basePrecision);
+			const price = new BN(order.price * pricePrecision);
```

Comment:
are you sure this is correct? confirming price/size isn't in lots here

Reply:
yeah observed it in tests and it looks right to me when divided by the same precisions

the price & size getters on `Order` call a fn to convert from lots to native

Reply:
```Bids
Price:  182.45
Size:  109.5904
Source:  { openbook: <BN: 198418cc00> }
Price:  182.43
Size:  136.984
Source:  { openbook: <BN: 1fe4e1f600> }
Price:  182.4
Size:  147.9472
Source:  { openbook: <BN: 2272571e00> }
Price:  182.38
Size:  180.8189
Source:  { openbook: <BN: 2a19a57420> }
Price:  182.33
Size:  164.3816
Source:  { openbook: <BN: 2645e82900> }
Price:  182.28
Size:  273.969
Source:  { openbook: <BN: 3fc9d32e40> }
Price:  174.08
Size:  30.3185
Asks
Price:  182.55
Size:  109.5624
Source:  { openbook: <BN: 19826d8d00> }
Price:  182.58
Size:  136.9497
Source:  { openbook: <BN: 1fe2d695a0> }
Price:  182.61
Size:  147.901
Source:  { openbook: <BN: 226f962940> }
Price:  182.63
Size:  180.7739
Source:  { openbook: <BN: 2a16f6cee0> }
Price:  182.68
Size:  164.3412
Source:  { openbook: <BN: 26437fb480> }
Price:  182.73
Size:  273.9029
Source:  { openbook: <BN: 3fc5e29320> }
```